lib/modules/datasource/julia-pkg-server/registry.ts

```diff
@@ -0,0 +1,247 @@
+import { Parse as tarParse } from 'tar';
+import { logger } from '../../../logger';
+import type { Http } from '../../../util/http';
+import { Result } from '../../../util/result';
+import { joinUrlParts } from '../../../util/url';
+import {
+  PKG_SERVER_REQUEST_HEADERS,
+  juliaPkgServerDatasourceId,
+} from './common';
+
+/**
+ * Generic container for aggregating and passing around the contents of files.
+ */
+interface FileContents<T extends Buffer[] | string> {
+  [path: string]: T;
+}
+
+/**
+ * Wraps all the information necessary to uniquely identify a registry at a
+ * particular point in time.
+ */
+interface JuliaPkgServerRegistry {
+  pkgServer: string;
+  state: string;
+  uuid: string;
+}
+
+/**
+ * URLs referencing a registry served by a PkgServer take the form:
+ * `<PkgServer host>/registry/<registry UUID>/<registry state>`.
+ *
+ * Although required when querying the state of a registry, the state in the
+ * URL is not mandatory when configuring a `registryUrl` passed into the
+ * datasource. Therefore that section is marked optional and should be ensured
+ * to be filled before further processing.
+ */
+const PKG_SERVER_REGISTRY_URL_FORMAT = new RegExp(
+  '^(?<pkgServer>.+)/registry/(?<uuid>\\p{Hex_Digit}{8}(-\\p{Hex_Digit}{4}){3}-\\p{Hex_Digit}{12})(/(?<state>\\p{Hex_Digit}{40}))?/?$',
+  'u',
+);
+
+/**
+ * The final part of the path of a (fully specified) registry URL references
+ * the tree SHA of the state of of the registry repository the package server
+ * is serving. Given the content-addressable nature of this SHA, it can be used
+ * as a caching mechanism.
+ *
+ * Cache keys are of the form `<registry UUID>:<registry state>`
+ */
+export function cacheKeyFromRegistryUrl(url: string): string {
+  return url.split('/').slice(-2).join(':');
+}
+
+// TODO: Cache this output as well?
+/**
+ * Extracts one or more files from the provided tarball in-memory.
+ *
+ * Returns `null`/fails if not all requested files can be extracted or if no
+ * files are requested for extraction.
+ */
+export async function extractFilesFromTarball(
+  tarball: Buffer,
+  pathsToExtract: string[],
+): Promise<FileContents<string> | null> {
+  if (pathsToExtract.length === 0) {
+    return null;
+  }
+
+  // Block until the content of all files has been (asynchronously) read from
+  // the tarball
+  return await new Promise<FileContents<string> | null>((resolve, reject) => {
+    // File content may be streamed in chunks which need to be aggregated
+    // TODO: The tests for this functionality only require a single "chunk" to
+    // be read from the tarball to succeed. Hence, it would be sufficient to
+    // just process a single chunk directly as a string and return it. This
+    // does not hold in general and a (minimal) test requiring multiple chunks
+    // should be implemented
```

Comment:
I have a tarball where this fails (which was obtained from the official `PkgServer), but that tarball is over 7MB large and I did not want to include that as a fixture. I have tried to create a more minimal example where the simpler approach fails, but have so far been unable to do so.

Reply:
How often do we need to fetch tarballs, if they can be large? And for how long can we cache them?

Reply:
That is a very good question and one I'm not entirely sure about myself.

The tarballs depend on the state of the registry, so they change whenever a new package is published to a registry. This is done by issuing a PR to the repository hosting the registry. So a suitable cache duration highly depends on the expected rate of package releases, when the corresponding PRs are merged, and how soon new releases should be available to Renovate. Given that these variables will very much depend on the registry from which dependencies are pulled, I'm not entirely sure there will be a uniformly suitable answer.

That being said, the `General` registry (which is where all public, open-source packages tend to be published) sees a couple of releases per hour and "automatically merges" suitable PRs every 8 minutes. So at a minimum a cache duration of 8 minutes can be used, although in practice it is probably fine to cache these up to an hour, maybe even a bit longer.

Reply:
If it's a single tarball per registry and not per package then I think something like a 15-30 minute cache time is fine. If it was common for the tarball to be unchanged for longer periods of time then we'd want to retain and use etag or last-modified headers for subsequent requests, but if registries typically get updated frequently then there's little gain from that